### ISSUE TITLE
cannot start container, System error write.

### ISSUE THREAD

**Issue by wallnerryan** [docker/swarm#1224]
When using Swarm to schedule container on a cluster, the error being seen is:
```
 [96] → docker run -it busybox sh
Error response from daemon: Cannot start container c65fa73e78fa0c3a1c830871f6b0fe8e7ac46504e4b1ee1a236b151ecce4dbe2: [8] System error: write /sys/fs/cgroup/memory/system.slice/docker-c65fa73e78fa0c3a1c830871f6b0fe8e7ac46504e4b1ee1a236b151ecce4dbe2.scope/memory.swappiness: invalid argument
```

Though I can run the docker command directly against the docker engines, (not swarm) with no problem. Which led me to file the issue here.
```
# docker run -it busybox sh
/ # 
/ # ls
bin      dev      etc      home     lib      lib64    linuxrc  media    mnt      opt      proc     root     run      sbin     sys      tmp      usr      var
/ # 
```

Swarm Servers Tried
```
Server:
 Version:      swarm/0.4.0
 API version:  1.16
 Go version:   go1.4.2
 Git commit:   d647d82
 Built:
 OS/Arch:      linux/amd64
```

```
Server:
 Version:      swarm/0.5.0-dev
 API version:  1.16
 Go version:   go1.5
 Git commit:   HEAD
 Built:
 OS/Arch:      linux/amd64
```

Docker Client
```
Client:
 Version:      1.8.2
 API version:  1.20
 Go version:   go1.4.2
 Git commit:   0a8c2e3
 Built:        Thu Sep 10 19:10:10 UTC 2015
 OS/Arch:      darwin/amd64
```

Docker Server
```
Server:
 Version:      1.8.0-dev
 API version:  1.20
 Go version:   go1.4.2
 Git commit:   5b92604
 Built:        Mon Jul 20 18:36:25 UTC 2015
 OS/Arch:      linux/amd64
 Experimental: true
```

Check-config on engine nodes
```
# sh check-config.sh 
warning: /proc/config.gz does not exist, searching other paths for kernel config ...
info: reading kernel config from /boot/config-3.10.0-123.el7.x86_64 ...

Generally Necessary:
- cgroup hierarchy: properly mounted [/sys/fs/cgroup]
- CONFIG_NAMESPACES: enabled
- CONFIG_NET_NS: enabled
- CONFIG_PID_NS: enabled
- CONFIG_IPC_NS: enabled
- CONFIG_UTS_NS: enabled
- CONFIG_DEVPTS_MULTIPLE_INSTANCES: enabled
- CONFIG_CGROUPS: enabled
- CONFIG_CGROUP_CPUACCT: enabled
- CONFIG_CGROUP_DEVICE: enabled
- CONFIG_CGROUP_FREEZER: enabled
- CONFIG_CGROUP_SCHED: enabled
- CONFIG_CPUSETS: enabled
- CONFIG_MEMCG: enabled
- CONFIG_MACVLAN: enabled (as module)
- CONFIG_VETH: enabled (as module)
- CONFIG_BRIDGE: enabled (as module)
- CONFIG_BRIDGE_NETFILTER: enabled
- CONFIG_NF_NAT_IPV4: enabled (as module)
- CONFIG_IP_NF_FILTER: enabled (as module)
- CONFIG_IP_NF_TARGET_MASQUERADE: enabled (as module)
- CONFIG_NETFILTER_XT_MATCH_ADDRTYPE: enabled (as module)
- CONFIG_NETFILTER_XT_MATCH_CONNTRACK: enabled (as module)
- CONFIG_NF_NAT: enabled (as module)
- CONFIG_NF_NAT_NEEDED: enabled
- CONFIG_POSIX_MQUEUE: enabled

Optional Features:
- CONFIG_MEMCG_KMEM: enabled
- CONFIG_MEMCG_SWAP: enabled
- CONFIG_MEMCG_SWAP_ENABLED: enabled
- CONFIG_RESOURCE_COUNTERS: enabled
- CONFIG_BLK_CGROUP: enabled
- CONFIG_IOSCHED_CFQ: enabled
- CONFIG_BLK_DEV_THROTTLING: enabled
- CONFIG_CGROUP_PERF: enabled
- CONFIG_CGROUP_HUGETLB: enabled
- CONFIG_NET_CLS_CGROUP: enabled
- CONFIG_NETPRIO_CGROUP: enabled (as module)
- CONFIG_CFS_BANDWIDTH: enabled
- CONFIG_FAIR_GROUP_SCHED: enabled
- CONFIG_RT_GROUP_SCHED: enabled
- CONFIG_EXT3_FS: missing
- CONFIG_EXT3_FS_XATTR: missing
- CONFIG_EXT3_FS_POSIX_ACL: missing
- CONFIG_EXT3_FS_SECURITY: missing
    (enable these ext3 configs if you are using ext3 as backing filesystem)
- CONFIG_EXT4_FS: enabled (as module)
- CONFIG_EXT4_FS_POSIX_ACL: enabled
- CONFIG_EXT4_FS_SECURITY: enabled
- Storage Drivers:
  - "aufs":
    - CONFIG_AUFS_FS: missing
  - "btrfs":
    - CONFIG_BTRFS_FS: enabled (as module)
  - "devicemapper":
    - CONFIG_BLK_DEV_DM: enabled (as module)
    - CONFIG_DM_THIN_PROVISIONING: enabled (as module)
  - "overlay":
    - CONFIG_OVERLAY_FS: missing
  - "zfs":
    - /dev/zfs: missing
    - zfs command: missing
    - zpool command: missing

```

**Comment by abronan** [comment]
Hi @wallnerryan, thanks for reporting the issue. What if you use another version of Docker, like 1.7.1, or 1.8.0 (not the dev binary)? Does this error still happen? Also what is the OS and the kernel version you are using? (we know there are issues with RHEL for example).

Would help if you can give us the output of `docker info` for the docker daemon and swarm :)

Thanks!

**Comment by wallnerryan** [comment]
@abronan 

I need the --volume-driver flag, so 1.8+ would be preferred. I can try 1.8.0, is the volume-driver flag in non experimental of 1.8.0?

Docker info on Daemon
```
# docker info
Containers: 6
Images: 15
Storage Driver: vfs
Execution Driver: native-0.2
Logging Driver: json-file
Kernel Version: 3.10.0-123.el7.x86_64
Operating System: CentOS Linux 7 (Core)
CPUs: 1
Total Memory: 1.798 GiB
Name: node2.local
ID: RF2O:Q3YS:WQSI:HXEH:55FH:L6Y5:Y5XO:IC5B:Q5HV:OO26:BSMC:A57Z
Experimental: true
```
Docker info on Swarm (mind the extra labels)
```
[117] → docker info
Containers: 9
Images: 10
Role: primary
Strategy: spread
Filters: affinity, health, constraint, port, dependency
Nodes: 3
 node1.local: 192.168.50.12:4243
  └ Containers: 1
  └ Reserved CPUs: 0 / 1
  └ Reserved Memory: 0 B / 1.888 GiB
  └ Labels: architecture=x86_64, bios_vendor=innotek GmbH, executiondriver=native-0.2, flocker/dataset=scaleio_flocker_driver, flocker/driver_capabilities=snapshot,consistency groups,thin provision,thick provision,read cache, is_virtual=True, kernelversion=3.10.0-123.el7.x86_64, memoryfree=1.14 GB, operatingsystem=CentOS Linux 7 (Core), physicalprocessorcount=1, selinux_enforced=False, storagedriver=vfs, swapfree=457.18 MB, system_load_10min=2.62, system_overload_10min=162
 node2.local: 192.168.50.13:4243
  └ Containers: 6
  └ Reserved CPUs: 0 / 1
  └ Reserved Memory: 0 B / 1.888 GiB
  └ Labels: architecture=x86_64, bios_vendor=innotek GmbH, executiondriver=native-0.2, flocker/dataset=scaleio_flocker_driver, flocker/driver_capabilities=snapshot,consistency groups,thin provision,thick provision,read cache, is_virtual=True, kernelversion=3.10.0-123.el7.x86_64, memoryfree=1.11 GB, operatingsystem=CentOS Linux 7 (Core), physicalprocessorcount=1, selinux_enforced=False, storagedriver=vfs, swapfree=417.95 MB, system_load_10min=2.71, system_overload_10min=171
 nod3.local: 192.168.50.11:4243
  └ Containers: 2
  └ Reserved CPUs: 0 / 1
  └ Reserved Memory: 0 B / 1.888 GiB
  └ Labels: architecture=x86_64, bios_vendor=innotek GmbH, executiondriver=native-0.2, flocker/dataset=scaleio_flocker_driver, flocker/driver_capabilities=snapshot,consistency groups,thin provision,thick provision,read cache, is_virtual=True, kernelversion=3.10.0-123.el7.x86_64, memoryfree=1.32 GB, operatingsystem=CentOS Linux 7 (Core), physicalprocessorcount=1, selinux_enforced=False, storagedriver=vfs, swapfree=545.36 MB, system_load_10min=4.16, system_overload_10min=316
CPUs: 3
Total Memory: 5.664 GiB
Name: node2.local
```

**Comment by abronan** [comment]
Thanks @wallnerryan, will try to see what's going on here but this seems related to the default value of `memory.swapiness` in `runc` for older kernels. I need to investigate and test it properly but I need a `CentOS` setup for that.

What would be worth a try is probably update your kernel to `3.18` and see if this gets rid of the issue. In the meantime we should still fix this for older kernels.

**Comment by wallnerryan** [comment]
@abronan would like to try and fix for this kernel, I have a dependency on this for current work. willing to dig in anywhere with some pointers

**Comment by wallnerryan** [comment]
@abronan since talking directly to the docker daemon works without any hiccup, where does the relevant code in swarm live that may try to set these limits on memory cgroups?

**Comment by wallnerryan** [comment]
@abronan  I see the ```ContainerConfig``` only has ```MemorySwap``` and setting this to -1 has no effect, is there a way to can set the ```Swappiness``` to -1? Not sure why it working via Docker CLI directly to the engine and not to swarm. I see in docker, the run config can set the cgroup to -1 which seems to be a solution for previous similar problems with runc.

**Comment by abronan** [comment]
@wallnerryan `Swapiness` is set to `-1` by default in `runc` (see opencontainers/runc#95), it should indeed work just fine using docker and older kernels. On the other side I don't recall that we have any `MemorySwapiness` parameter in `dockerclient` so this could be the issue of the remote client being out of sync with the update that was introduced in docker `1.8` and the remote API endpoint of the daemon not initializing it to `-1` in this case.

**Comment by wallnerryan** [comment]
Yes I did see that, so I thought it should work just fine too. If I use the cli on the hose that swarm would choose e.g "constraint:node==node1" then the container comes up no problem. So this would tell me it's an issue with Swarm and whatever request it's passing might be out of sync? that make sense? 

This still wouldn't make sense to me (im less familiar with swarm code base though) since swarm just creates the ContainerConfig from dockerclient which doesn't do anything special to swappiness. So not sure where in the pipeline this may be getting set but it's between DockerClient on the host, or why it's trying to be written to. If I check the swappiness on the host, it's set to 30, and o cannot change its value, but docker doesn't Compton about this, just swarm. Could it be possibly docker engine not setting this value on init? I am running 1.8 but it may be a dev cut when volume driver was first introduced.

-Ryan

**Comment by wallnerryan** [comment]
changed the dockerclient on the host which was ```1.8.0-dev``` to ```1.8.0``` and this is fixed. :| sorry for taking this issue on a run around but i would assume the dev release didnt treat swappiness on centos correct. Seems fixed for now so I'm going to close this.

**Comment by abronan** [comment]
@wallnerryan Yes you probably used a `1.8.0-dev` without the runC patch for `MemorySwapiness`. Glad this is fixed with `1.8.0`. Thanks for the update :+1:
